File: backend/api/language_schedules.py
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from backend.database import get_db
from backend.models.language_schedule import LanguageSchedule
from backend.schemas.language_schedule import (
    LanguageScheduleResponse,
    LanguageScheduleUpsert,
)

router = APIRouter(prefix="/api/language-schedules", tags=["language_schedules"])


@router.get("", response_model=list[LanguageScheduleResponse])
def list_schedules(db: Session = Depends(get_db)):
    return (
        db.query(LanguageSchedule)
        .order_by(LanguageSchedule.language_code.asc())
        .all()
    )


@router.put("/{lang}", response_model=LanguageScheduleResponse)
def upsert_schedule(
    lang: str, body: LanguageScheduleUpsert, db: Session = Depends(get_db)
):
    existing = (
        db.query(LanguageSchedule)
        .filter(LanguageSchedule.language_code == lang)
        .first()
    )
    if existing:
        existing.time_brt = body.time_brt
        existing.updated_at = datetime.utcnow()
        row = existing
    else:
        row = LanguageSchedule(
            language_code=lang,
            time_brt=body.time_brt,
            updated_at=datetime.utcnow(),
        )
        db.add(row)
    db.commit()
    db.refresh(row)
    return row


@router.delete("/{lang}", status_code=204)
def delete_schedule(lang: str, db: Session = Depends(get_db)):
    existing = (
        db.query(LanguageSchedule)
        .filter(LanguageSchedule.language_code == lang)
        .first()
    )
    if not existing:
        raise HTTPException(404, "Schedule not found")
    db.delete(existing)
    db.commit()
    return None

File: frontend/src/api/languageSchedules.ts
import type { LanguageSchedule } from '../types'

const BASE = '/api/language-schedules'

export async function listSchedules(): Promise<LanguageSchedule[]> {
  const r = await fetch(BASE)
  if (!r.ok) throw new Error('Failed to load schedules')
  return r.json()
}

export async function upsertSchedule(lang: string, time_brt: string): Promise<LanguageSchedule> {
  const r = await fetch(`${BASE}/${encodeURIComponent(lang)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ time_brt }),
  })
  if (!r.ok) throw new Error('Failed to upsert schedule')
  return r.json()
}

export async function deleteSchedule(lang: string): Promise<void> {
  const r = await fetch(`${BASE}/${encodeURIComponent(lang)}`, { method: 'DELETE' })
  if (!r.ok) throw new Error('Failed to delete schedule')
}
